File: 01_basics/01_10_palindrome_2/Palindrome.java
public class Palindrome {
    public static void main(String[] args) {
        String st = "A man, a plan, a canal: Panama";
        System.out.println(st);
        System.out.println(String.format("Is Palindrome: %b", isPalindrome(st)));
    }

    private static boolean isPalindrome(String s) {
        int i = 0;
        int j = s.length() - 1;

        while (i < j) {
            while (!isAlNum(s.charAt(i))) {
                i++;
            }
            while (!isAlNum(s.charAt(j))) {
                j--;
            }
            if (s.toLowerCase().charAt(i++) != s.toLowerCase().charAt(j--)) {
                return false;
            }
        }

        return true;
    }

    private static boolean isAlNum(char ch) {
        return ('a' <= ch && ch <= 'z') || ('A' <= ch && ch <= 'Z') || ('0' <= ch && ch <= '9');
    }
}

File: 01_basics/01_10_palindrome_2/palindrome.js
function isPalindrome(s) {
    let i = 0;
    let j = s.length - 1;

    while (i < j) {
        while (i < j && !isAlphanumeric(s[i])) {
            i++;
        }
        while (i < j && !isAlphanumeric(s[j])) {
            j--;
        }
        if (i < j && s[i].toLowerCase() !== s[j].toLowerCase()) {
            return false;
        }
        i++;
        j--;
    }

    return true;
}

function isAlphanumeric(char) {
    const code = char.charCodeAt(0);
    return (code >= '0'.charCodeAt(0) && code <= '9'.charCodeAt(0)) ||
        (code >= 'A'.charCodeAt(0) && code <= 'Z'.charCodeAt(0)) ||
        (code >= 'a'.charCodeAt(0) && code <= 'z'.charCodeAt(0));
}

const str = "A man, a plan, a canal: Panama";
console.log(str);
console.log("Is Palindrome:", isPalindrome(str));